### PR TITLE
Refs [TASK][26259] Init MVP structure

### PR DESCRIPTION
**Pull request by toandv98**
## Related Tickets

- [#Task 26259](https://edu-redmine.sun-asterisk.vn/issues/26259)

## WHAT

- Init MVP structure

## HOW

- Create BaseActivity, BaseFragment, BaseView, BasePresenter

## WHY

- move common methods across child to parent, stop declaring it in each child

## Review Checklist

| Category      | View Point                                                   | Description                                                  | Expected Reviewer Answer | Reviewer1 (ToanDV) | Reviewer2 (HuyTQ)   |
| ------------- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------ | ------------------ | ------------------ |
| General       | Code should be self-explanatory.                             | Get a feel of story reading, while going through the code. Use appropriate name for variables, functions and classes. If you are taking more time to understand the code, then either code needs refactoring or at least comments have to be written to make it clear. | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| General       | Is this function or class too big? If yes, is the function or class having too many responsibilities? |                                                              | NO                       | <li>- [x] no</li>  | <li>- [x] no</li>  |
| General       | Does the code follow DRY?: Do not Repeat Yourself            | Is there any redundant or duplicate code? DRY (Do not Repeat Yourself) principle: The same code should not be repeated more than twice. Consider reusable services, functions and components. Consider generic functions and classes. | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| General       | Exception handling and cleanup (dispose) resources           |                                                              | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| General       | Does the code build and execute correctly?                   |                                                              | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| Conventions   | Does the source code have end-line errors?                   |                                                              | NO                       | <li>- [x] no</li>  | <li>- [x] no</li>  |
| Conventions   | No hard coding, use constants/configuration values           |                                                              | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| Conventions   | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| Redmine       | Does the ticket follow Sun* Redmine working process?         | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |                                                              | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |

## Notes (Optional)

*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/edu/sun/oereminder/ui/base/MvpView.kt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package com.edu.sun.oereminder.ui.base
+
+import androidx.annotation.StringRes
+
+interface MvpView {
```
Comment (datdt-1935) [inline app/src/main/java/com/edu/sun/oereminder/ui/base/MvpView.kt]:
```suggestion
interface BaseView {
```

File: app/src/main/java/com/edu/sun/oereminder/ui/base/MvpPresenter.kt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package com.edu.sun.oereminder.ui.base
+
+interface MvpPresenter<V : MvpView> {
```
Comment (datdt-1935) [inline app/src/main/java/com/edu/sun/oereminder/ui/base/MvpPresenter.kt]:
```suggestion
interface BasePresenter<V : BaseView> {
```

File: app/src/main/java/com/edu/sun/oereminder/ui/base/BaseActivity.kt [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    abstract fun initListener()
+
+    override fun onDestroy() {
+        super.onDestroy()
+        presenter?.detachView()
+    }
+
+    override fun showMessage(msg: String) {
+        Toast.makeText(this, msg, Toast.LENGTH_SHORT).show()
```
Comment (huytq-2422) [inline app/src/main/java/com/edu/sun/oereminder/ui/base/BaseActivity.kt]: viết trong extension của Context nhé 
Tạo file Context+.kt

File: app/src/main/java/com/edu/sun/oereminder/ui/base/BaseActivity.kt [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    override fun showMessage(resId: Int) {
+        showMessage(getString(resId))
```
Comment (huytq-2422) [inline app/src/main/java/com/edu/sun/oereminder/ui/base/BaseActivity.kt]: tương tự

File: app/src/main/res/layout/activity_main.xml [hunk 5]
```diff
@@ -4,7 +4,7 @@
     xmlns:tools="http://schemas.android.com/tools"
     android:layout_width="match_parent"
     android:layout_height="match_parent"
-    tools:context=".MainActivity">
+    tools:context=".ui.main.MainActivity">
 
     <TextView
```
Comment (tuannm-1276) [inline app/src/main/res/layout/activity_main.xml:7]: remove

File: app/src/main/java/com/edu/sun/oereminder/ui/base/BaseActivity.kt (hunk 3 above)
Comment (toandv98) [reply app/src/main/java/com/edu/sun/oereminder/ui/base/BaseActivity.kt]: Đây là function của BaseView để gọi nó trong Presenter. Nếu dùng extension thì làm như thế nào nhỉ

Comment (huytq-2422) [reply app/src/main/java/com/edu/sun/oereminder/ui/base/BaseActivity.kt]: viết extension để viết gọn lại hàm Toast dùng nhiều lần về sau

**Pull request merged**